File: typescript/src/slip10.ts
// SLIP-0010 hierarchical derivation for the ed25519 curve (hardened-only) — the
// derivation Ledger uses for Ed25519 accounts.

import { concatBytes } from "@noble/curves/utils.js";
import { hmacSha512 } from "./primitives.ts";

const HARDENED = 0x80000000;

/** A SLIP-0010 node. */
export interface Node {
  privateKey: Uint8Array;
  chainCode: Uint8Array;
}

/** Master node from a BIP-39 (or any) seed. */
export function master(seed: Uint8Array): Node {
  return split(hmacSha512(utf8("ed25519 seed"), seed));
}

/** One hardened child step (the hardened bit is added automatically). */
export function deriveChild(parent: Node, index: number): Node {
  const idx = new Uint8Array(4);
  new DataView(idx.buffer).setUint32(0, index + HARDENED, false);
  return split(hmacSha512(parent.chainCode, concatBytes(Uint8Array.of(0), parent.privateKey, idx)));
}

/** Derive along a path such as "m/44'/9381'/0'/0'/0'". Every level is hardened. */
export function derivePath(seed: Uint8Array, path: string): Node {
  let node = master(seed);
  for (const index of parsePath(path)) {
    node = deriveChild(node, index);
  }
  return node;
}

export function parsePath(path: string): number[] {
  const t = path.trim();
  if (t !== "m" && !t.startsWith("m/")) {
    throw new Error(`path must start with 'm': ${path}`);
  }
  if (t === "m") {
    return [];
  }
  return t
    .slice(2)
    .split("/")
    .map((raw) => {
      const cleaned = raw.replace(/['hH]+$/, "");
      if (!/^\d+$/.test(cleaned)) {
        throw new Error(`bad path segment: '${raw}'`);
      }
      const n = Number(cleaned);
      if (n >= HARDENED) {
        throw new Error(`index out of range: ${raw}`);
      }
      return n;
    });
}

function split(i: Uint8Array): Node {
  return { privateKey: i.slice(0, 32), chainCode: i.slice(32, 64) };
}

function utf8(s: string): Uint8Array {
  return new TextEncoder().encode(s);
}

File: rust/src/slip10.rs
//! SLIP-0010 hierarchical derivation for the ed25519 curve (hardened-only) —
//! the derivation Ledger uses for Ed25519 accounts.

use crate::Error;
use crate::primitives::hmac_sha512;

const HARDENED: u32 = 0x8000_0000;

/// A SLIP-0010 node.
#[derive(Clone)]
pub struct Node {
    pub private_key: [u8; 32],
    pub chain_code: [u8; 32],
}

/// Master node from a BIP-39 (or any) seed.
pub fn master(seed: &[u8]) -> Node {
    split(hmac_sha512(b"ed25519 seed", seed))
}

/// One hardened child step (the hardened bit is added automatically).
pub fn derive_child(parent: &Node, index: u32) -> Node {
    let hardened = index | HARDENED;
    let mut data = Vec::with_capacity(1 + 32 + 4);
    data.push(0u8);
    data.extend_from_slice(&parent.private_key);
    data.extend_from_slice(&hardened.to_be_bytes());
    split(hmac_sha512(&parent.chain_code, &data))
}

/// Derive along a path such as `m/44'/9381'/0'/0'/0'`. Every level is hardened.
pub fn derive_path(seed: &[u8], path: &str) -> Result<Node, Error> {
    let mut node = master(seed);
    for index in parse_path(path)? {
        node = derive_child(&node, index);
    }
    Ok(node)
}

pub fn parse_path(path: &str) -> Result<Vec<u32>, Error> {
    let trimmed = path.trim();
    if trimmed != "m" && !trimmed.starts_with("m/") {
        return Err(Error::Path(format!("path must start with 'm': {path}")));
    }
    if trimmed == "m" {
        return Ok(vec![]);
    }
    trimmed[2..]
        .split('/')
        .map(|raw| {
            let cleaned = raw.trim_end_matches(['\'', 'h', 'H']);
            let n: u64 = cleaned
                .parse()
                .map_err(|_| Error::Path(format!("bad path segment: '{raw}'")))?;
            if n >= HARDENED as u64 {
                return Err(Error::Path(format!("index out of range: {raw}")));
            }
            Ok(n as u32)
        })
        .collect()
}

fn split(i: [u8; 64]) -> Node {
    let mut private_key = [0u8; 32];
    let mut chain_code = [0u8; 32];
    private_key.copy_from_slice(&i[..32]);
    chain_code.copy_from_slice(&i[32..]);
    Node {
        private_key,
        chain_code,
    }
}
